                                                                                        Weijia Yu &Yuxuan Zhang                                                                          

# Iris Dataset classification exercise 

## Problem Statement

This dataset was created by R.A. Fisher and good for pattern recognition. For this exercise, we will use this dataset to test three classification models: KNN,Navie Bayes, Logistic regression. We will rank and discuss the model performance from best to worst

In [62]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_iris

from sklearn.model_selection import train_test_split

from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score



from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_validate
from sklearn.model_selection import ShuffleSplit


from sklearn.metrics import classification_report, confusion_matrix

# This means my results look the same each time 
np.random.seed(47)

## Data Wrangling and Describe

In [63]:
iris=load_iris() #load iris help classify iris dataset specifically, this function is from sklearn
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [111]:
print(iris.DESCR) # prints the description of the dataset

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

This shows a satistic rundown of the whole dataset, including mean, min, max and std etc.

In [106]:
print(iris.target)
#we want to take a look ot what the target looks like, return shows the three classes for classification

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


In [105]:
iris.target_names
#returned array shows the name for each class


array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

The function returns five categories of output. The ones we are using are going to be data and target. data is an array of each flowers'length and width, target is the three categories of flower type, indicated by number 0-2

## KNN Classification 

In [66]:
#store each in a different variable 
X=iris.data
y=iris.target #three catergories

In [67]:
import warnings
warnings.filterwarnings('ignore')

When taking a look at the dataset, the data are very nicely ordered therefore I want to use shufflesplit to randomly split the test and train data, which is going to be used in KNN model, I want to set the iteration nmc=10.

In [68]:
#setting up the KNN classifier 
nmc=10
knn = KNeighborsClassifier(n_neighbors=2)
# scikit-learn tools
shuffle = ShuffleSplit(n_splits=nmc, test_size=0.3) 
#since iris dataset there are 3 features, I want to do 70/30 on train/test data

CVInfo = cross_validate(knn, X, y, cv=shuffle,return_train_score=True)
print(np.mean(CVInfo['train_score']))
print(np.mean(CVInfo['test_score']))



0.9790476190476192
0.9377777777777778


In [69]:
# Fit the model on the full dataset and get predictions
knn.fit(X, y)
y_predknn = knn.predict(X)

KNN model can be evaluated through confusion matrix on its accuarcy, precision, recall to see its performance

In [70]:
#set the confususion matric
cm = confusion_matrix(y, y_predknn)
cr = classification_report(y, y_predknn)

In [71]:
# Print report
print("Confusion matrix:\n", cm)
print("Classification report:\n", cr)

Confusion matrix:
 [[50  0  0]
 [ 0 50  0]
 [ 0  3 47]]
Classification report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        50
           1       0.94      1.00      0.97        50
           2       1.00      0.94      0.97        50

    accuracy                           0.98       150
   macro avg       0.98      0.98      0.98       150
weighted avg       0.98      0.98      0.98       150



Looking at the confusion matrix. since the dataset has three features (0,1,2). model made mistake on the third feature which is 3 false negative predictions and the rest are good.
precision, recall, and f1 across all three classes have good scores. The one we want to pay attention to is the recall which is true positive(when the class is predicted correctly). the accuracy of the model is also 0.98 which means the model predicted 98% of the data.


## Naive Bayes classification

In [81]:
np.random.seed(42)

# set up Naive Bayes classifier
NBC = GaussianNB()
nmc=10
# Again, scikit-learn tools
shuffle = ShuffleSplit(n_splits=nmc, test_size=.5)
CVInfo = cross_validate(NBC, X, y, cv=shuffle,return_train_score=True)
print(np.mean(CVInfo['train_score']))
print(np.mean(CVInfo['test_score']))




0.9626666666666667
0.9453333333333334


In [82]:
# Fit the model on the full dataset and get predictions
NBC.fit(X, y)
y_predNBC = NBC.predict(X)

In [86]:
#set the confususion matric
cm = confusion_matrix(y, y_predNBC)
cr = classification_report(y, y_predNBC)

In [87]:
# Print report
print("Confusion matrix:\n", cm)
print("Classification report:\n", cr)

Confusion matrix:
 [[50  0  0]
 [ 0 47  3]
 [ 0  3 47]]
Classification report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        50
           1       0.94      0.94      0.94        50
           2       0.94      0.94      0.94        50

    accuracy                           0.96       150
   macro avg       0.96      0.96      0.96       150
weighted avg       0.96      0.96      0.96       150



Looking at the confusion matrix. since the dataset has three features (0,1,2). model made mistake on the third feature which is 3 false negative predictions as well as the second feature for 3 false negative predictions as well. 
Precision, recall, and f1 across all three classes have a good score, however, only the first feature has the perfect precision, recall, and F1 score. recall rate is 0.96 for both the second and third features.
For the NBC model, the accuracy is 0.96.

## Logistic Regression classification

In [98]:
# set up Logistic Regression classifier
lr = LogisticRegression()
#set up a iteration of 2000 time using this model



In [99]:
# set up Naive Bayes classifier

nmc=10
# Again, scikit-learn tools
shuffle = ShuffleSplit(n_splits=nmc, test_size=.5)
CVInfo = cross_validate(lr, X, y, cv=shuffle,return_train_score=True)
print(np.mean(CVInfo['train_score']))
print(np.mean(CVInfo['test_score']))

0.9733333333333334
0.9626666666666667


In [100]:
lr.fit(X, y)
y_predlr = lr.predict(X)

In [104]:
#set the confususion matric
cm = confusion_matrix(y, y_predlr)
cr = classification_report(y, y_predlr)


In [102]:
# Print report
print("Confusion matrix:\n", cm)
print("Classification report:\n", cr)

Confusion matrix:
 [[50  0  0]
 [ 0 47  3]
 [ 0  1 49]]
Classification report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        50
           1       0.98      0.94      0.96        50
           2       0.94      0.98      0.96        50

    accuracy                           0.97       150
   macro avg       0.97      0.97      0.97       150
weighted avg       0.97      0.97      0.97       150



For logistic model, it has a better confusion matrix score as the third feature has one false negative instance compared to the NBC model. 

As the same before, we want to pay more attention to accuracy which is 0.97 in this model, however even third class has a higher recall rate which is 0.98, it has a lower precision rate which is 0.94.

# Conclusion


As shown from the previous analyses, I think it's safe to rank these three model from worse to best this way:

NBC <- logistic <- kNN based on its overall accuracy 